File: NowPlaying/iPhone/Classes/AbstractDVDBlurayCache.m
// Copyright 2008 Cyrus Najmabadi
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#import "AbstractDVDBlurayCache.h"

#import "Application.h"
#import "DVD.h"
#import "DateUtilities.h"
#import "FileUtilities.h"
#import "ImageUtilities.h"
#import "LargePosterCache.h"
#import "LinkedSet.h"
#import "Movie.h"
#import "NetworkUtilities.h"
#import "NowPlayingAppDelegate.h"
#import "NowPlayingModel.h"
#import "PointerSet.h"
#import "ThreadingUtilities.h"
#import "Utilities.h"
#import "XmlElement.h"

@interface AbstractDVDBlurayCache()
@property (retain) PointerSet* moviesSetData;
@property (retain) NSArray* moviesData;
@property (retain) LinkedSet* prioritizedMovies;
@property (retain) NSMutableDictionary* bookmarksData;
@end


@implementation AbstractDVDBlurayCache

@synthesize moviesSetData;
@synthesize moviesData;
@synthesize prioritizedMovies;
@synthesize bookmarksData;

- (void) dealloc {
    self.moviesSetData = nil;
    self.moviesData = nil;
    self.prioritizedMovies = nil;
    self.bookmarksData = nil;

    [super dealloc];
}


- (id) initWithModel:(NowPlayingModel*) model_ {
    if (self = [super initWithModel:model_]) {
        self.prioritizedMovies = [LinkedSet set];
    }

    return self;
}


- (NSString*) serverAddress {
    @throw [NSException exceptionWithName:@"ImproperSubclassing" reason:@"" userInfo:nil];
}


- (NSString*) directory {
    @throw [NSException exceptionWithName:@"ImproperSubclassing" reason:@"" userInfo:nil];
}


- (NSString*) detailsDirectory {
    return [self.directory stringByAppendingPathComponent:@"Details"];
}


- (NSString*) postersDirectory {
    return [self.directory stringByAppendingPathComponent:@"Posters"];
}


- (NSString*) imdbDirectory {
    return [self.directory stringByAppendingPathComponent:@"IMDb"];
}


- (NSString*) moviesFile {
    return [[self directory] stringByAppendingPathComponent:@"Movies.plist"];
}


- (NSString*) bookmarksFile {
    return [[self directory] stringByAppendingPathComponent:@"Bookmarks.plist"];
}


- (NSArray*) decodeMovieArray:(NSArray*) array {
    if (array == nil) {
        return [NSArray array];
    }

    NSMutableArray* result = [NSMutableArray array];
    for (NSDictionary* dictionary in array) {
        [result addObject:[Movie movieWithDictionary:dictionary]];
    }

    return result;
}


- (NSArray*) loadMovies:(NSString*) file {
    NSArray* encodedMovies = [FileUtilities readObject:file];
    return [self decodeMovieArray:encodedMovies];
}


- (NSArray*) loadMovies {
    return [self loadMovies:self.moviesFile];
}


- (void) setMovies:(NSArray*) array {
    self.moviesData = array;
    self.moviesSetData = [PointerSet setWithArray:array];
}


- (NSArray*) movies {
    if (moviesData == nil) {
        [self setMovies:[self loadMovies]];
    }

    return moviesData;
}


- (PointerSet*) moviesSet {
    [self movies];
    return moviesSetData;
}


- (NSMutableDictionary*) loadBookmarks {
    NSArray* movies = [self loadMovies:self.bookmarksFile];
    if (movies.count == 0) {
        return [NSMutableDictionary dictionary];
    }

    NSMutableDictionary* result = [NSMutableDictionary dictionary];
    for (Movie* movie in movies) {
        [result setObject:movie forKey:movie.canonicalTitle];
    }

    return result;
}


- (NSMutableDictionary*) bookmarks {
    if (bookmarksData == nil) {
        self.bookmarksData = [self loadBookmarks];
    }

    return bookmarksData;
}


- (void) updateMovies {
    [ThreadingUtilities performSelector:@selector(updateMoviesBackgroundEntryPoint)
                               onTarget:self
               inBackgroundWithGate:gate
                                visible:YES];
}


- (void) updateDetails {
    [ThreadingUtilities performSelector:@selector(updateDetailsBackgroundEntryPoint:)
                               onTarget:self
               inBackgroundWithArgument:self.movies
                                   gate:gate
                                visible:NO];
}


- (void) update {
    if (model.userAddress.length == 0) {
        return;
    }

    [self updateMovies];
    [self updateDetails];
}


- (NSArray*) split:(NSString*) value {
    if (value.length == 0) {
        return [NSArray array];
    }

    return [value componentsSeparatedByString:@"/"];
}


- (NSString*) massage:(NSString*) text {
    unichar a1[] = { 0xE2, 0x20AC, 0x201C };
    text = [text stringByReplacingOccurrencesOfString:[NSString stringWithCharacters:a1 length:ArrayLength(a1)]
                                           withString:@"-"];

    unichar a2[] = { 0xEF, 0xBF, 0xBD };
    text = [text stringByReplacingOccurrencesOfString:[NSString stringWithCharacters:a2 length:ArrayLength(a2)]
                                           withString:@"'"];

    unichar a3[] = { 0xE2, 0x20AC, 0x153 };
    text = [text stringByReplacingOccurrencesOfString:[NSString stringWithCharacters:a3 length:ArrayLength(a3)]
                                           withString:@"\""];

    unichar a4[] = { 0xE2, 0x20AC, 0x9D };
    text = [text stringByReplacingOccurrencesOfString:[NSString stringWithCharacters:a4 length:ArrayLength(a4)]
                                           withString:@"\""];

    unichar a5[] = { 0xE2, 0x20AC, 0x2122 };
    text = [text stringByReplacingOccurrencesOfString:[NSString stringWithCharacters:a5 length:ArrayLength(a5)]
                                           withString:@"'"];

    unichar a6[] = { 0xC2, 0xA0 };
    text = [text stringByReplacingOccurrencesOfString:[NSString stringWithCharacters:a6 length:ArrayLength(a6)]
                                           withString:@" "];

    unichar a7[] = { 0xE2, 0x20AC, 0x201D };
    text = [text stringByReplacingOccurrencesOfString:[NSString stringWithCharacters:a7 length:ArrayLength(a7)]
                                           withString:@"-"];

    unichar a8[] = { 0xC2, 0xAE };
    text = [text stringByReplacingOccurrencesOfString:[NSString stringWithCharacters:a8 length:ArrayLength(a8)]
                                           withString:@"®"];

    unichar a9[] = { 0xE2, 0x20AC, 0xA2 };
    text = [text stringByReplacingOccurrencesOfString:[NSString stringWithCharacters:a9 length:ArrayLength(a9)]
                                           withString:@"•"];

    return text;
}


- (void) processVideoElement:(XmlElement*) videoElement
                      result:(NSMutableDictionary*) result {
    NSString* title = [videoElement attributeValue:@"title"];
    NSString* releaseDateString = [videoElement attributeValue:@"release_date"];
    NSString* price = [videoElement attributeValue:@"retail_price"];
    NSString* rating = [videoElement attributeValue:@"mpaa_rating"];
    NSString* format = [videoElement attributeValue:@"format"];
    NSArray* genres = [self split:[videoElement attributeValue:@"genre"]];
    NSArray* cast = [self split:[videoElement attributeValue:@"cast"]];
    NSArray* directors = [self split:[videoElement attributeValue:@"director"]];
    NSString* discs = [videoElement attributeValue:@"discs"];
    NSString* poster = [videoElement attributeValue:@"image"];
    NSString* synopsis = [videoElement attributeValue:@"synopsis"];
    NSDate* releaseDate = [DateUtilities parseIS08601Date:releaseDateString];
    NSString* url = [videoElement attributeValue:@"url"];
    NSString* length = [videoElement attributeValue:@"length"];
    NSString* studio = [videoElement attributeValue:@"studio"];

    synopsis = [self massage:synopsis];

    DVD* dvd = [DVD dvdWithTitle:title
                           price:price
                          format:format
                           discs:discs
                             url:url];

    Movie* movie = [Movie movieWithIdentifier:[NSString stringWithFormat:@"%d", dvd]
                                        title:title
                                       rating:rating
                                       length:[length intValue]
                                  releaseDate:releaseDate
                                  imdbAddress:@""
                                       poster:poster
                                     synopsis:synopsis
                                       studio:studio
                                    directors:directors
                                         cast:cast
                                       genres:genres];

    [result setObject:dvd forKey:movie];
}


- (NSDictionary*) processElement:(XmlElement*) element {
    NSMutableDictionary* result = [NSMutableDictionary dictionary];

    for (XmlElement* child in element.children) {
        NSAutoreleasePool* pool = [[NSAutoreleasePool alloc] init];
        {
            [self processVideoElement:child result:result];
        }
        [pool release];
    }

    return result;
}


- (NSString*) detailsFile:(Movie*) movie set:(PointerSet*) movies {
    if (movies == nil || [movies containsObject:movie]) {
        return [[[self detailsDirectory] stringByAppendingPathComponent:[FileUtilities sanitizeFileName:movie.canonicalTitle]]
                stringByAppendingString:@".plist"];
    }

    return nil;
}


- (NSString*) detailsFile:(Movie*) movie {
    NSAssert([NSThread isMainThread], @"");

    return [self detailsFile:movie set:self.moviesSet];
}


- (NSString*) imdbFile:(Movie*) movie set:(PointerSet*) movies {
    if (movies == nil || [movies containsObject:movie]) {
        return [[[self imdbDirectory] stringByAppendingPathComponent:[FileUtilities sanitizeFileName:movie.canonicalTitle]]
                stringByAppendingString:@".plist"];
    }

    return nil;
}


- (NSString*) imdbFile:(Movie*) movie {
    NSAssert([NSThread isMainThread], @"");

    return [self imdbFile:movie set:self.moviesSet];
}


- (NSString*) posterFile:(Movie*) movie set:(PointerSet*) movies {
    if (movies == nil || [movies containsObject:movie]) {
        return [[[self postersDirectory] stringByAppendingPathComponent:[FileUtilities sanitizeFileName:movie.canonicalTitle]]
                stringByAppendingString:@".jpg"];
    }

    return nil;
}


- (NSString*) smallPosterFile:(Movie*) movie set:(PointerSet*) movies {
    if (movies == nil || [movies containsObject:movie]) {
        return [[[self postersDirectory] stringByAppendingPathComponent:[FileUtilities sanitizeFileName:movie.canonicalTitle]]
                stringByAppendingString:@"-small.png"];
    }

    return nil;
}


- (NSString*) posterFile:(Movie*) movie {
    NSAssert([NSThread isMainThread], @"");

    return [self posterFile:movie set:self.moviesSet];
}


- (NSString*) smallPosterFile:(Movie*) movie {
    NSAssert([NSThread isMainThread], @"");

    return [self smallPosterFile:movie set:self.moviesSet];
}


- (void) saveVideosArray:(NSArray*) videos toFile:(NSString*) file {
    NSMutableArray* encoded = [NSMutableArray array];
    for (Movie* video in videos) {
        [encoded addObject:video.dictionary];
    }
    [FileUtilities writeObject:encoded toFile:file];
}


- (void) saveData:(NSDictionary*) dictionary {
    NSArray* videos = dictionary.allKeys;

    for (Movie* movie in dictionary) {
        DVD* dvd = [dictionary objectForKey:movie];
        [FileUtilities writeObject:dvd.dictionary toFile:[self detailsFile:movie set:nil]];
    }

    // do this last.  it signifies that we're done
    [self saveVideosArray:videos toFile:self.moviesFile];
}


- (NSSet*) cachedDirectoriesToClear {
    return [NSSet setWithObjects:
            [self detailsDirectory],
            [self imdbDirectory],
            [self postersDirectory], nil];
}


- (void) updateMoviesBackgroundEntryPoint {
    NSDate* lastUpdateDate = [FileUtilities modificationDate:self.moviesFile];
    if (lastUpdateDate != nil) {
        if (ABS(lastUpdateDate.timeIntervalSinceNow) < (3 * ONE_DAY)) {
            return;
        }
    }

    XmlElement* element = [NetworkUtilities xmlWithContentsOfAddress:self.serverAddress
                                                           important:YES];

    if (element == nil) {
        return;
    }

    NSDictionary* map = [self processElement:element];

    if (map.count == 0) {
        return;
    }

    [self saveData:map];

    [self performSelectorOnMainThread:@selector(reportResults:)
                           withObject:map
                        waitUntilDone:NO];
}


- (void) saveBookmarks {
    [self saveVideosArray:self.bookmarks.allValues toFile:self.bookmarksFile];
}


- (void) reportResults:(NSDictionary*) map {
    NSMutableArray* movies = [NSMutableArray arrayWithArray:map.allKeys];
    // add in any previously bookmarked movies that we now no longer know about.
    for (Movie* movie in self.bookmarks.allValues) {
        if (![movies containsObject:movie]) {
            [movies addObject:movie];
        }
    }

    // also determine if any of the data we found match items the user bookmarked
    for (Movie* movie in movies) {
        if ([model isBookmarked:movie]) {
            [self.bookmarks setObject:movie forKey:movie.canonicalTitle];
        }
    }
    [self saveBookmarks];

    [self setMovies:movies];
    [self updateDetails];
    [NowPlayingAppDelegate majorRefresh];
}


- (void) updateVideoPoster:(Movie*) movie {
    if (movie == nil) {
        return;
    }

    NSString* file = [self posterFile:movie set:nil];
    if ([FileUtilities fileExists:file]) {
        return;
    }

    if (movie.poster.length == 0) {
        [model.largePosterCache downloadFirstPosterForMovie:movie];
        return;
    }

    NSString* address = [NSString stringWithFormat:@"http://%@.appspot.com/LookupCachedResource?q=%@",
                         [Application host], [Utilities stringByAddingPercentEscapes:movie.poster]];

    NSData* data = [NetworkUtilities dataWithContentsOfAddress:address important:NO];
    if (data != nil) {
        [FileUtilities writeData:data toFile:file];
        [NowPlayingAppDelegate minorRefresh];
    }
}


- (void) updateVideoImdbAddress:(Movie*) movie {
    if (movie == nil) {
        return;
    }

    NSString* imdbFile = [self imdbFile:movie set:nil];

    NSDate* lastLookupDate = [FileUtilities modificationDate:imdbFile];
    if (lastLookupDate != nil) {
        NSString* value = [FileUtilities readObject:imdbFile];
        if (value.length > 0) {
            // we have a real imdb value for this movie
            return;
        }

        // we have a sentinel.  only update if it's been long enough
        if (ABS(lastLookupDate.timeIntervalSinceNow) < (3 * ONE_DAY)) {
            return;
        }
    }

    NSString* url = [NSString stringWithFormat:@"http://%@.appspot.com/LookupIMDbListings?q=%@", [Application host], [Utilities stringByAddingPercentEscapes:movie.canonicalTitle]];
    NSString* imdbAddress = [NetworkUtilities stringWithContentsOfAddress:url important:NO];
    if (imdbAddress == nil) {
        return;
    }

    // write down the response (even if it is empty).  An empty value will
    // ensure that we don't update this entry too often.
    [FileUtilities writeObject:imdbAddress toFile:imdbFile];
    if (imdbAddress.length > 0) {
        [NowPlayingAppDelegate minorRefresh];
    }
}


- (Movie*) getNextMovie:(NSMutableArray*) movies {
    Movie* movie = [prioritizedMovies removeLastObjectAdded];

    if (movie != nil) {
        return movie;
    }

    if (movies.count > 0) {
        movie = [[[movies lastObject] retain] autorelease];
        [movies removeLastObject];
        return movie;
    }

    return nil;
}


- (void) prioritizeMovie:(Movie*) movie {
    if (![self.moviesSet containsObject:movie]) {
        return;
    }

    [prioritizedMovies addObject:movie];
}


- (void) updateDetailsBackgroundEntryPoint:(NSArray*) movies {
    NSMutableArray* videos = [NSMutableArray arrayWithArray:movies];

    Movie* movie;
    do {
        NSAutoreleasePool* pool = [[NSAutoreleasePool alloc] init];
        {
            movie = [self getNextMovie:videos];
            [self updateVideoPoster:movie];
            [self updateVideoImdbAddress:movie];
        }
        [pool release];
    } while (movie != nil);
}


- (NSString*) imdbAddressForMovie:(Movie*) movie {
    return [FileUtilities readObject:[self imdbFile:movie]];
}


- (UIImage*) posterForMovie:(Movie*) movie {
    return [UIImage imageWithData:[FileUtilities readData:[self posterFile:movie]]];
}


- (UIImage*) smallPosterForMovie:(Movie*) movie {
    NSString* smallPosterPath = [self smallPosterFile:movie];
    NSData* smallPosterData;

    if ([FileUtilities size:smallPosterPath] == 0) {
        NSData* normalPosterData = [FileUtilities readData:[self posterFile:movie]];
        smallPosterData = [ImageUtilities scaleImageData:normalPosterData
                                                toHeight:SMALL_POSTER_HEIGHT];

        [FileUtilities writeData:smallPosterData
                          toFile:smallPosterPath];
    } else {
        smallPosterData = [FileUtilities readData:smallPosterPath];
    }

    return [UIImage imageWithData:smallPosterData];
}


- (DVD*) detailsForMovie:(Movie*) movie {
    NSDictionary* dictionary = [FileUtilities readObject:[self detailsFile:movie]];
    if (dictionary == nil) {
        return nil;
    }

    return [DVD dvdWithDictionary:dictionary];
}


- (void) addBookmark:(NSString*) canonicalTitle {
    for (Movie* movie in self.movies) {
        if ([movie.canonicalTitle isEqual:canonicalTitle]) {
            [self.bookmarks setObject:movie forKey:canonicalTitle];
            [self saveBookmarks];
            return;
        }
    }
}


- (void) removeBookmark:(NSString*) canonicalTitle {
    [self.bookmarks removeObjectForKey:canonicalTitle];
    [self saveBookmarks];
}

@end

File: NowPlaying/Android/src/org/metasyntactic/caches/dvd/AbstractDVDBlurayCache.java
package org.metasyntactic.caches.dvd;

import org.metasyntactic.NowPlayingApplication;
import org.metasyntactic.NowPlayingModel;
import org.metasyntactic.caches.AbstractMovieCache;
import org.metasyntactic.collections.IdentityHashSet;
import org.metasyntactic.data.DVD;
import org.metasyntactic.data.Movie;
import org.metasyntactic.threading.ThreadingUtilities;
import org.metasyntactic.utilities.CollectionUtilities;
import org.metasyntactic.utilities.DateUtilities;
import org.metasyntactic.utilities.FileUtilities;
import org.metasyntactic.utilities.NetworkUtilities;
import static org.metasyntactic.utilities.StringUtilities.isNullOrEmpty;
import static org.metasyntactic.utilities.XmlUtilities.children;
import org.w3c.dom.Element;
import org.w3c.dom.Node;

import java.io.File;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.Collections;
import java.util.Date;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;

public abstract class AbstractDVDBlurayCache extends AbstractMovieCache {
  private Set<Movie> moviesSet;
  private List<Movie> movies;

  private boolean updated;

  protected AbstractDVDBlurayCache(final NowPlayingModel model) {
    super(model);
    getDirectory().mkdirs();
    getDetailsDirectory().mkdirs();
  }

  protected abstract String getServerAddress();

  protected abstract File getDirectory();

  private File getDetailsDirectory() {
    return new File(getDirectory(), "Details");
  }

  private File getMoviesFile() {
    return new File(getDirectory(), "Movies");
  }

  @Override protected List<File> getCacheDirectories() {
    return Arrays.asList(getDirectory(), getDetailsDirectory());
  }

  private static List<Movie> loadMovies(final File file) {
    final List<Movie> movies = FileUtilities.readPersistableList(Movie.reader, file);
    if (CollectionUtilities.isEmpty(movies)) {
      return Collections.emptyList();
    }

    return movies;
  }

  private List<Movie> loadMovies() {
    return loadMovies(getMoviesFile());
  }

  private void setMovies(final List<Movie> movies) {
    this.movies = movies;
    moviesSet = new IdentityHashSet<Movie>(movies);
  }

  private Object getMovies() {
    if (movies == null) {
      setMovies(loadMovies());
    }

    return movies;
  }

  private Collection<Movie> getMoviesSet() {
    getMovies();
    return moviesSet;
  }

  public void update() {
    if (isNullOrEmpty(model.getUserAddress())) {
      return;
    }

    if (updated) {
      return;
    }
    updated = true;

    final String name = getClass().getSimpleName() + "-Update";
    final Runnable runnable = new Runnable() {
      public void run() {
        updateBackgroundEntryPoint();
      }
    };
    ThreadingUtilities.performOnBackgroundThread(name, runnable, null, true);
  }

  private static List<String> split(final String value) {
    if (isNullOrEmpty(value)) {
      return Collections.emptyList();
    }

    return Arrays.asList(value.split("/"));
  }

  private static final Map<String, String> mapping = new LinkedHashMap<String, String>();

  static {
    mapping.put(new String(new char[]{0xE2, 0x20AC, 0x201C}), "-");
    mapping.put(new String(new char[]{0xEF, 0xBF, 0xBD}), "'");
    mapping.put(new String(new char[]{0xE2, 0x20AC, 0x153}), "\"");
    mapping.put(new String(new char[]{0xE2, 0x20AC, 0x9D}), "\"");
    mapping.put(new String(new char[]{0xE2, 0x20AC, 0x2122}), "'");
    mapping.put(new String(new char[]{0xC2, 0xA0}), " ");
    mapping.put(new String(new char[]{0xE2, 0x20AC, 0x201D}), "-");
    mapping.put(new String(new char[]{0xC2, 0xAE}), "®");
    mapping.put(new String(new char[]{0xE2, 0x20AC, 0xA2}), "•");
  }

  private static String massage(String text) {
    for (final Map.Entry<String, String> entry : mapping.entrySet()) {
      text = text.replace(entry.getKey(), entry.getValue());
    }

    return text;
  }

  private static int identifier = Integer.MAX_VALUE;

  private static void processVideoElement(final Element videoElement, final Map<Movie, DVD> result) {
    final String title = videoElement.getAttribute("title");
    final String releaseDateString = videoElement.getAttribute("release_date");
    final String price = videoElement.getAttribute("retail_price");
    final String rating = videoElement.getAttribute("mpaa_rating");
    final String format = videoElement.getAttribute("format");
    final List<String> genres = split(videoElement.getAttribute("genre"));
    final List<String> cast = split(videoElement.getAttribute("cast"));
    final List<String> directors = split(videoElement.getAttribute("directors"));
    final String discs = videoElement.getAttribute("discs");
    final String poster = videoElement.getAttribute("image");
    final String synopsis = massage(videoElement.getAttribute("synopsis"));

    final Date releaseDate = DateUtilities.parseISO8601Date(releaseDateString);
    final String url = videoElement.getAttribute("url");
    final String length = videoElement.getAttribute("length");
    final String studio = videoElement.getAttribute("studio");

    int intLength;
    try {
      intLength = Integer.parseInt(length);
    } catch (final NumberFormatException ignored) {
      intLength = 0;
    }

    final DVD dvd = new DVD(title, price, format, discs, url);
    final Movie movie = new Movie(String.valueOf(identifier), title, rating, intLength, "", releaseDate, poster, synopsis, studio, directors, cast,
        genres);
    identifier--;

    result.put(movie, dvd);
  }

  private Map<Movie, DVD> processElement(final Node element) {
    final Map<Movie, DVD> result = new HashMap<Movie, DVD>();

    for (final Element child : children(element)) {
      if (shutdown) { return null; }
      processVideoElement(child, result);
    }

    return result;
  }

  private File getDetailsFile(final Movie movie, final Collection<Movie> movies) {
    if (movies == null || movies.contains(movie)) {
      return new File(getDetailsDirectory(), FileUtilities.sanitizeFileName(movie.getCanonicalTitle()));
    }

    return null;
  }

  private File getDetailsFile(final Movie movie) {
    return getDetailsFile(movie, getMoviesSet());
  }

  private void saveData(final Map<Movie, DVD> map) {
    for (final Map.Entry<Movie, DVD> entry : map.entrySet()) {
      if (shutdown) { return; }
      FileUtilities.writePersistable(entry.getValue(), getDetailsFile(entry.getKey(), null));
    }

    // do this last.  it signifies that we're done
    FileUtilities.writePersistableCollection(map.keySet(), getMoviesFile());
  }

  private List<Movie> updateBackgroundEntryPointWorker() {
    final File moviesFile = getMoviesFile();
    if (moviesFile.exists()) {
      if (FileUtilities.daysSinceNow(moviesFile) < 3) {
        return null;
      }
    }

    final Element element = NetworkUtilities.downloadXml(getServerAddress(), true);
    if (element == null) {
      return null;
    }

    if (shutdown) { return null; }
    final Map<Movie, DVD> map = processElement(element);
    if (CollectionUtilities.isEmpty(map)) {
      return null;
    }

    if (shutdown) { return null; }
    saveData(map);
    updatedMoviesClear();

    final Runnable runnable = new Runnable() {
      public void run() {
        reportResults(map);
      }
    };
    ThreadingUtilities.performOnMainThread(runnable);
    return new ArrayList<Movie>(map.keySet());
  }


  private void updateBackgroundEntryPoint() {
    List<Movie> localMovies = updateBackgroundEntryPointWorker();
    if (shutdown) { return; }
    if (CollectionUtilities.isEmpty(localMovies)) {
      localMovies = loadMovies();
    }

    addPrimaryMovies(localMovies);
  }

  private void reportResults(final Map<Movie, DVD> map) {
    setMovies(new ArrayList<Movie>(map.keySet()));
    NowPlayingApplication.refresh();
  }

  private void updatePoster(final Movie movie) {
    model.getPosterCache().update(movie);
  }

  private void updateIMDb(final Movie movie) {
    model.getIMDbCache().update(movie);
  }

  /*
  private void updateAmazon(final Movie movie) {

  }

  private void updateWikipedia(final Movie movie) {

  }
   */

  @Override public void prioritizeMovie(final Movie movie) {
    if (!getMoviesSet().contains(movie)) {
      return;
    }

    super.prioritizeMovie(movie);
  }

  @Override protected void updateMovieDetails(final Movie movie) {
    if (shutdown) { return; }
    updatePoster(movie);
    if (shutdown) { return; }
    updateIMDb(movie);
    //updateWikipedia(movie);
    //updateAmazon(movie);
  }

  public DVD detailsForMovie(final Movie movie) {
    return FileUtilities.readPersistable(DVD.reader, getDetailsFile(movie));
  }
}